File: apps/medousa-home/src/lib/types/workflow.ts
import type { RecurringRunEntry } from "$lib/types/recurring";

export type WorkflowStepKind = "grapheme" | "prompt" | "mcp";

export interface WorkflowStepGrapheme {
  kind: "grapheme";
  id: string;
  source: string;
}

export interface WorkflowStepPrompt {
  kind: "prompt";
  id: string;
  user_prompt: string;
  system_prompt?: string | null;
}

export interface WorkflowStepMcp {
  kind: "mcp";
  id: string;
  server_id: string;
  tool_name: string;
  args?: Record<string, unknown>;
  effect_class?: string | null;
}

export type WorkflowStepSpec =
  | WorkflowStepGrapheme
  | WorkflowStepPrompt
  | WorkflowStepMcp;

export interface WorkflowRunRequest {
  name?: string | null;
  strategy?: string;
  mode?: string;
  steps: WorkflowStepSpec[];
  on_failure?: string;
  note?: string | null;
  queue?: string | null;
}

export interface WorkflowListEntry {
  workflow_id: string;
  name?: string | null;
  status: string;
  strategy: string;
  mode: string;
  root_job_id: string;
  root_job_state?: string | null;
  scheduled_recurring_id?: string | null;
  created_at_utc: string;
  step_count: number;
}

export interface WorkflowsListResponse {
  count: number;
  workflows: WorkflowListEntry[];
}

export interface WorkflowStepResultDto {
  id: string;
  kind: string;
  status: string;
  output?: unknown;
  error?: string | null;
}

export interface WorkflowDetailResponse {
  workflow_id: string;
  name?: string | null;
  status: string;
  strategy: string;
  mode: string;
  on_failure: string;
  note?: string | null;
  root_job_id: string;
  root_job_state?: string | null;
  scheduled_recurring_id?: string | null;
  created_at_utc: string;
  steps: WorkflowStepSpec[];
  step_results: WorkflowStepResultDto[];
}

export interface WorkflowRunResponse {
  workflow_id: string;
  status: string;
  strategy: string;
  root_job_id: string;
  job_type: string;
  lane: string;
}

export interface WorkflowPlanRequest {
  goal: string;
  context?: Record<string, unknown> | null;
}

export interface WorkflowScheduleSuggestion {
  cron_expr: string;
  timezone: string;
}

export interface WorkflowPlanResponse {
  goal: string;
  confidence: string;
  execute_with: string;
  suggested_workflow?: WorkflowRunRequest | null;
  suggested_schedule?: WorkflowScheduleSuggestion | null;
  suggested_tool_input?: unknown;
  notes: string[];
  assumptions: string[];
}

export interface WorkflowScheduleRequest extends WorkflowRunRequest {
  cron_expr: string;
  timezone?: string;
  display_name?: string | null;
  recurring_id?: string | null;
  delivery?: Record<string, unknown> | null;
  enabled?: boolean;
}

export interface WorkflowScheduleResponse {
  workflow_id: string;
  status: string;
  recurring_id: string;
  cron_expr: string;
  timezone: string;
  next_run_at_utc: string;
  materialized_job_id?: string | null;
}

export interface WorkflowRunsResponse {
  workflow_id: string;
  count: number;
  runs: RecurringRunEntry[];
}

export interface FlowComposerDraft {
  name: string;
  goal: string;
  steps: WorkflowStepSpec[];
  cron_expr: string;
  timezone: string;
}

export function newStepId(prefix: string): string {
  return `${prefix}-${crypto.randomUUID().slice(0, 8)}`;
}

export function emptyFlowDraft(): FlowComposerDraft {
  return {
    name: "",
    goal: "",
    steps: [],
    cron_expr: "0 9 * * *",
    timezone: "UTC",
  };
}

export function workflowRunRequestFromDraft(
  draft: FlowComposerDraft,
): WorkflowRunRequest {
  return {
    name: draft.name.trim() || null,
    strategy: "sequential",
    mode: "default",
    on_failure: "stop",
    steps: draft.steps,
    queue: "default",
  };
}

File: crates/medousa-types/src/workflow.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum WorkflowStepSpec {
    Grapheme {
        id: String,
        source: String,
    },
    Prompt {
        id: String,
        user_prompt: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        system_prompt: Option<String>,
    },
    Mcp {
        id: String,
        server_id: String,
        tool_name: String,
        #[serde(default)]
        args: Value,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        effect_class: Option<String>,
    },
    ToolReplay {
        id: String,
        tool_name: String,
        #[serde(default)]
        input: Value,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        session_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        slice_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        tool_round: Option<usize>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        run_id: Option<String>,
        #[serde(default)]
        requires_confirm: bool,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowRunRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default = "default_strategy")]
    pub strategy: String,
    #[serde(default = "default_mode")]
    pub mode: String,
    pub steps: Vec<WorkflowStepSpec>,
    #[serde(default = "default_on_failure")]
    pub on_failure: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub note: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub queue: Option<String>,
}

fn default_strategy() -> String {
    "sequential".to_string()
}

fn default_mode() -> String {
    "default".to_string()
}

fn default_on_failure() -> String {
    "stop".to_string()
}

impl WorkflowStepSpec {
    pub fn id(&self) -> &str {
        match self {
            Self::Grapheme { id, .. } | Self::Prompt { id, .. } | Self::Mcp { id, .. }
            | Self::ToolReplay { id, .. } => id,
        }
    }
}
